subsys/usb/host/class/usbh_hub_mgr.c

```diff
@@ -0,0 +1,1243 @@
+		for (int i = 0; i < hub_mgr_data->num_ports; i++) {
+			/* Initialize port instance */
+			hub_mgr_data->port_list[i].udev = NULL;
+			hub_mgr_data->port_list[i].reset_count = 3;
+			hub_mgr_data->port_list[i].port_status = kPortRunWaitPortChange;
+			hub_mgr_data->port_list[i].state = PORT_STATE_DISCONNECTED;
+			hub_mgr_data->port_list[i].port_num = i + 1;
+			hub_mgr_data->port_list[i].speed = USB_SPEED_SPEED_FS; /* Default full speed */
+		}
+		
+		hub_mgr_data->hub_status = kHubRunIdle;
+		hub_mgr_data->state = HUB_STATE_OPERATIONAL;
+		need_prime_interrupt = 1;
+		LOG_INF("Hub initialization completed, starting interrupt monitoring");
+		break;
+
+	case kHubRunGetStatusDone:
+		hub_mgr_data->hub_status = kHubRunClearDone;
+		LOG_DBG("Hub status get completed");
+		need_prime_interrupt = 1;
+		process_success = 1;
+		break;
+
+	case kHubRunClearDone:
+		hub_mgr_data->hub_status = kHubRunIdle;
+		need_prime_interrupt = 1;
+		process_success = 1;
+		break;
+
+	default:
+		LOG_ERR("Unknown hub status: %d", hub_mgr_data->hub_status);
+		hub_mgr_data->hub_status = kHubRunInvalid;
+		hub_mgr_data->state = HUB_STATE_ERROR;
+		break;
+	}
+
+	k_mutex_unlock(&hub_mgr_data->lock);
+
+	if (need_prime_interrupt) {
+		hub_mgr_data->hub_status = kHubRunIdle;
+		if (!hub_mgr_data->int_active && !hub_mgr_data->being_removed) {
+			ret = usbh_hub_mgr_start_interrupt(hub_mgr_data);
+			if (ret) {
+				LOG_ERR("Failed to start interrupt monitoring: %d", ret);
+			}
+		}
+	} else if (!process_success && hub_mgr_data->hub_status != kHubRunInvalid) {
+		hub_mgr_data->hub_status = kHubRunInvalid;
+		hub_mgr_data->state = HUB_STATE_ERROR;
+	}
+}
+
+/**
+ * @brief Recursively disconnect hub and all children
+ */
+static void usbh_hub_mgr_recursive_disconnect(struct usbh_hub_mgr_data *hub_mgr_data)
+{
+	if (!hub_mgr_data) {
+		return;
+	}
+
+	k_mutex_lock(&hub_mgr_data->lock, K_FOREVER);
+
+	if (hub_mgr_data->being_removed) {
+		k_mutex_unlock(&hub_mgr_data->lock);
+		LOG_DBG("Hub level %d already being removed", hub_mgr_data->hub_level);
+		return;
+	}
+
+	hub_mgr_data->being_removed = true;
+
+	LOG_INF("Recursively disconnecting Hub level %d and all children", 
+		hub_mgr_data->hub_level);
+
+	k_work_cancel_delayable(&hub_mgr_data->port_work);
+	k_work_cancel_delayable(&hub_mgr_data->hub_work);
+	hub_mgr_data->int_active = false;
+
+	if (hub_mgr.current_hub_process == hub_mgr_data) {
+		hub_mgr.current_hub_process = NULL;
+		hub_mgr_data->current_port = 0;
+	}
+
+	for (int i = 0; i < hub_mgr_data->num_ports; i++) {
+		if (hub_mgr_data->port_list && hub_mgr_data->port_list[i].udev) {
+			struct usb_device *port_udev = hub_mgr_data->port_list[i].udev;
+			struct usbh_hub_mgr_data *child_hub = find_hub_mgr_by_udev(port_udev);
+			if (child_hub) {
+				LOG_INF("Found child Hub on port %d, recursing", i + 1);
+				k_mutex_unlock(&hub_mgr_data->lock);
+				
+				usbh_hub_mgr_recursive_disconnect(child_hub);
+				
+				k_mutex_lock(&hub_mgr_data->lock, K_FOREVER);
```

Comment:
It is supported to lock a mutex multiple time in Zephyr.

They just need to be unlocked as many time as they are locked.

Would that allow to remove the `unlock()`/`lock()` used here?